File: backend/src/main/java/org/luxons/sevenwonders/game/wonders/WonderStage.java
package org.luxons.sevenwonders.game.wonders;

import java.util.List;

import org.luxons.sevenwonders.game.api.Table;
import org.luxons.sevenwonders.game.boards.Board;
import org.luxons.sevenwonders.game.cards.CardBack;
import org.luxons.sevenwonders.game.cards.Requirements;
import org.luxons.sevenwonders.game.effects.Effect;
import org.luxons.sevenwonders.game.resources.BoughtResources;

public class WonderStage {

    private Requirements requirements;

    private List<Effect> effects;

    private CardBack cardBack;

    public Requirements getRequirements() {
        return requirements;
    }

    public void setRequirements(Requirements requirements) {
        this.requirements = requirements;
    }

    public List<Effect> getEffects() {
        return effects;
    }

    public void setEffects(List<Effect> effects) {
        this.effects = effects;
    }

    public boolean isBuilt() {
        return cardBack != null;
    }

    public boolean isBuildable(Table table, int playerIndex, List<BoughtResources> boughtResources) {
        Board board = table.getBoard(playerIndex);
        return requirements.areMetWithHelpBy(board, boughtResources);
    }

    void build(CardBack cardBack) {
        this.cardBack = cardBack;
    }

    void activate(Table table, int playerIndex, List<BoughtResources> boughtResources) {
        effects.forEach(e -> e.apply(table, playerIndex));
    }
}

File: sw-engine/src/main/kotlin/org/luxons/sevenwonders/engine/wonders/WonderStage.kt
package org.luxons.sevenwonders.engine.wonders

import org.luxons.sevenwonders.engine.Player
import org.luxons.sevenwonders.engine.boards.Board
import org.luxons.sevenwonders.engine.cards.Requirements
import org.luxons.sevenwonders.engine.effects.Effect
import org.luxons.sevenwonders.model.cards.CardBack
import org.luxons.sevenwonders.model.resources.ResourceTransactions

internal class WonderStage(
    val requirements: Requirements,
    val effects: List<Effect>
) {
    var cardBack: CardBack? = null
        private set

    val isBuilt: Boolean
        get() = cardBack != null

    fun isBuildable(board: Board, boughtResources: ResourceTransactions): Boolean =
        requirements.areMetWithHelpBy(board, boughtResources)

    fun placeCard(cardBack: CardBack) {
        this.cardBack = cardBack
    }

    fun activate(player: Player, boughtResources: ResourceTransactions) {
        effects.forEach { it.applyTo(player) }
        requirements.pay(player, boughtResources)
    }
}
